File: frontend/src/lib/request.ts
import type { ClaimKind, ProofRequest } from './data/types.ts'
import { fromBase64Url, hex, toBase64Url, utf8 } from './pof/bytes.ts'

// A proof request lives entirely in the URL. Nothing is stored anywhere.
// Mirrors backend/crates/pof-prove/src/request.rs rule for rule; fixtures/requests.json holds
// the links both sides must agree on.

const KINDS: ClaimKind[] = ['HoldsAtLeast', 'HoldsExactly', 'ReceivedPayment', 'ReceivedAtLeastSince']
/** The claim kinds a v1 prover can answer. The others are on the roadmap. */
export const PROVABLE: ClaimKind[] = ['HoldsAtLeast']
export const ZAT_PER_UNIT = 12_500_000n
export const MAX_ZATOSHI = 21_000_000n * 100_000_000n
const FIELDS = new Set(['v', 'id', 'claim', 'zatoshi', 'audience', 'expiryDays', 'respondBy', 'bind'])

export function newRequestId(): string {
  const b = new Uint8Array(8)
  crypto.getRandomValues(b)
  return hex(b)
}

/**
 * The problem with an audience identifier, or null. Printable ASCII only, so trimming and
 * lower-casing (inside the audience hash) mean exactly the same thing here and in Rust.
 */
export function audienceProblem(id: string): string | null {
  if (!/^[\x20-\x7e]*$/.test(id)) return 'The audience may use only printable ASCII: letters, digits, spaces and punctuation.'
  if (!id.trim()) return 'Name the one party this proof is for.'
  return null
}

export function encodeRequest(r: ProofRequest): string {
  return toBase64Url(utf8(JSON.stringify(r)))
}

export type DecodedRequest = { ok: true; request: ProofRequest } | { ok: false; reason: string }

const fail = (reason: string): DecodedRequest => ({ ok: false, reason })

export function decodeRequestDetailed(s: string): DecodedRequest {
  // Links wrapped by mail clients pick up whitespace; anything else outside base64url is damage.
  const clean = s.replace(/[\t\n\f\r ]/g, '')
  const bytes = /^[A-Za-z0-9_-]*$/.test(clean) ? fromBase64Url(clean) : null
  if (!bytes) return fail('The link is not valid base64url. It may have been cut off when it was copied.')
  let raw: unknown
  try {
    // fatal + ignoreBOM: invalid UTF-8 and a leading BOM are refused, as serde_json refuses them
    raw = JSON.parse(new TextDecoder('utf-8', { fatal: true, ignoreBOM: true }).decode(bytes))
  } catch {
    return fail('The request does not parse.')
  }
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) return fail('The request does not parse.')
  const unknown = Object.keys(raw).find((k) => !FIELDS.has(k))
  if (unknown !== undefined) return fail(`Unknown request field "${unknown}".`)
  // An explicit null is the same as leaving the field out.
  const r = Object.fromEntries(Object.entries(raw).filter(([, v]) => v !== null)) as Record<string, unknown>
  if (r.v !== 1) return fail('Unknown request version.')
  if (typeof r.claim !== 'string' || !KINDS.includes(r.claim as ClaimKind)) return fail('Unknown claim kind.')
  if (!PROVABLE.includes(r.claim as ClaimKind)) return fail(`This prover cannot answer ${r.claim} requests yet.`)
  if (typeof r.zatoshi !== 'string' || !/^[0-9]+$/.test(r.zatoshi)) return fail('The amount is missing or malformed.')
  const zat = BigInt(r.zatoshi)
  if (zat === 0n || zat > MAX_ZATOSHI) return fail('The amount is out of range.')
  if (zat % ZAT_PER_UNIT !== 0n) return fail('Thresholds are proven in 0.125 ZEC steps; this one is not a multiple of 0.125.')
  if (typeof r.audience !== 'string') return fail('The request does not name who it is for.')
  const audience = audienceProblem(r.audience)
  if (audience) return fail(audience)
  if (typeof r.expiryDays !== 'number' || !Number.isInteger(r.expiryDays) || r.expiryDays < 1 || r.expiryDays > 365) return fail('Expiry must be 1–365 days.')
  if (r.id !== undefined && (typeof r.id !== 'string' || !/^[0-9a-f]{16}$/.test(r.id))) return fail('Malformed request id.')
  if (r.respondBy !== undefined && (typeof r.respondBy !== 'number' || !Number.isSafeInteger(r.respondBy) || r.respondBy < 1)) return fail('Malformed response deadline.')
  if (r.bind !== undefined && r.bind !== 'solana') return fail('Unknown binding.')
  return { ok: true, request: r as unknown as ProofRequest }
}

// Both commands assume the working directory the READMEs use: backend/.

/** For holders with their own wallet. Keys never leave their machine. The snapshot path is where `pof-anchor scan` writes by default. */
export function cliCommand(encoded: string, bind?: 'solana') {
  return `pof-prove prove --request ${encoded} --snapshot target/snapshots/mainnet-<height>.vsnp --seed-file ~/.vouch/seed.txt${bind === 'solana' ? ' --bind-solana <your-solana-pubkey>' : ''} --out proof.pof`
}

/** The same request answered by the demo holder on the demo ledger. */
export function demoCliCommand(encoded: string, bind?: 'solana') {
  return `pof-prove demo prove --world ../fixtures/demo-world.json --request ${encoded}${bind === 'solana' ? ' --bind-solana <your-solana-pubkey>' : ''} --out proof.pof`
}

File: backend/crates/pof-prove/src/request.rs
//! A proof request lives entirely in a URL: base64url(JSON). Mirrors `frontend/src/lib/request.ts`.

use serde::{Deserialize, Serialize};

/// Largest `respond_by` both sides represent exactly: JavaScript's `Number.MAX_SAFE_INTEGER`.
pub const MAX_RESPOND_BY: u64 = (1 << 53) - 1;

/// Unknown fields are refused, and an explicit JSON `null` in an optional field reads as
/// absent, exactly as on the web.
#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ProofRequest {
    pub v: u8,
    /// Random request id, 16 lowercase hex. Lets a verifier match a response to its request.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    pub claim: String,
    /// Zatoshi as a decimal string, so large values survive JSON.
    #[serde(with = "dec")]
    pub zatoshi: u64,
    pub audience: String,
    pub expiry_days: u32,
    /// Unix seconds the verifier wants an answer by.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub respond_by: Option<u64>,
    /// "solana" when the proof must be bound to the holder's Solana account (on-chain audiences).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub bind: Option<String>,
}

mod dec {
    use serde::{de::Error, Deserialize, Deserializer, Serializer};
    pub fn serialize<S: Serializer>(v: &u64, s: S) -> Result<S::Ok, S::Error> {
        s.serialize_str(&v.to_string())
    }
    /// ASCII digits only: `u64::from_str` alone would also take a leading `+`.
    pub fn deserialize<'de, D: Deserializer<'de>>(d: D) -> Result<u64, D::Error> {
        let s = String::deserialize(d)?;
        if s.is_empty() || !s.bytes().all(|b| b.is_ascii_digit()) {
            return Err(D::Error::custom("zatoshi must be a decimal string"));
        }
        s.parse().map_err(D::Error::custom)
    }
}

#[derive(Debug, thiserror::Error)]
pub enum RequestError {
    #[error("the request is not valid base64url")]
    Encoding,
    #[error("the request does not parse: {0}")]
    Json(String),
    #[error("{0}")]
    Invalid(&'static str),
}

/// The problem with an audience identifier, if any. Printable ASCII only, so trimming and
/// lower-casing (inside the audience hash) mean exactly the same thing here and in JavaScript.
pub fn audience_problem(id: &str) -> Option<&'static str> {
    if !id.bytes().all(|b| (0x20..=0x7e).contains(&b)) {
        return Some("the audience may use only printable ASCII: letters, digits, spaces and punctuation");
    }
    id.trim().is_empty().then_some("the request names no audience")
}

impl ProofRequest {
    /// The `r=` value of a request link. Accepts exactly what `decodeRequestDetailed` in
    /// `frontend/src/lib/request.ts` accepts; `fixtures/requests.json` holds the cases.
    pub fn decode(encoded: &str) -> Result<Self, RequestError> {
        // Links wrapped by mail clients pick up whitespace; anything else outside base64url is damage.
        let clean: String = encoded.chars().filter(|c| !c.is_ascii_whitespace()).collect();
        if !clean.bytes().all(|b| b.is_ascii_alphanumeric() || b == b'-' || b == b'_') {
            return Err(RequestError::Encoding);
        }
        let bytes = pof_core::from_base64url(&clean).map_err(|_| RequestError::Encoding)?;
        // Through a Value, so duplicate keys resolve last-wins as in JSON.parse, and only an
        // object is a request (serde would also build the struct from an array).
        let value: serde_json::Value = serde_json::from_slice(&bytes).map_err(|e| RequestError::Json(e.to_string()))?;
        if !value.is_object() {
            return Err(RequestError::Json("a request is a JSON object".into()));
        }
        let r: ProofRequest = serde_json::from_value(value).map_err(|e| RequestError::Json(e.to_string()))?;
        r.validate()?;
        Ok(r)
    }

    pub fn validate(&self) -> Result<(), RequestError> {
        if self.v != 1 {
            return Err(RequestError::Invalid("unknown request version"));
        }
        if self.claim != "HoldsAtLeast" {
            return Err(RequestError::Invalid("this prover answers HoldsAtLeast requests"));
        }
        if self.zatoshi == 0 || self.zatoshi > pof_core::MAX_ZATOSHI {
            return Err(RequestError::Invalid("the threshold is out of range"));
        }
        if !self.zatoshi.is_multiple_of(pof_core::ZAT_PER_UNIT) {
            return Err(RequestError::Invalid("the threshold must be a multiple of 0.125 ZEC"));
        }
        if let Some(p) = audience_problem(&self.audience) {
            return Err(RequestError::Invalid(p));
        }
        if !(1..=365).contains(&self.expiry_days) {
            return Err(RequestError::Invalid("expiry must be 1–365 days"));
        }
        if self.id.as_ref().is_some_and(|id| id.len() != 16 || !id.bytes().all(|b| matches!(b, b'0'..=b'9' | b'a'..=b'f'))) {
            return Err(RequestError::Invalid("malformed request id"));
        }
        if self.respond_by.is_some_and(|t| !(1..=MAX_RESPOND_BY).contains(&t)) {
            return Err(RequestError::Invalid("malformed response deadline"));
        }
        if self.bind.as_deref().is_some_and(|b| b != "solana") {
            return Err(RequestError::Invalid("unknown binding"));
        }
        Ok(())
    }

    /// The claim as an English sentence, identical to the web review screen.
    pub fn sentence(&self) -> String {
        format!(
            "{} asks you to prove you hold at least {} ZEC. Valid {} day{} after you generate it.",
            self.audience.trim(),
            zec(self.zatoshi),
            self.expiry_days,
            if self.expiry_days == 1 { "" } else { "s" }
        )
    }
}

pub fn zec(z: u64) -> String {
    let whole = z / 100_000_000;
    let frac = format!("{:08}", z % 100_000_000);
    let frac = frac.trim_end_matches('0');
    let mut w = String::new();
    let s = whole.to_string();
    for (i, c) in s.chars().enumerate() {
        if i > 0 && (s.len() - i).is_multiple_of(3) {
            w.push(',');
        }
        w.push(c);
    }
    if frac.is_empty() {
        w
    } else {
        format!("{w}.{frac}")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The same links `frontend/scripts/verify-fixtures.ts` checks: both sides must agree.
    #[test]
    fn shared_request_vectors() {
        let path = concat!(env!("CARGO_MANIFEST_DIR"), "/../../../fixtures/requests.json");
        let file: serde_json::Value = serde_json::from_slice(&std::fs::read(path).unwrap()).unwrap();
        let cases = file["cases"].as_array().unwrap();
        assert!(cases.len() > 50);
        for c in cases {
            let (name, r, ok) = (c["name"].as_str().unwrap(), c["r"].as_str().unwrap(), c["ok"].as_bool().unwrap());
            if let Some(json) = c["json"].as_str() {
                assert_eq!(pof_core::from_base64url(r).unwrap(), json.as_bytes(), "{name}: json and r disagree");
            }
            let got = ProofRequest::decode(r);
            assert_eq!(got.is_ok(), ok, "{name}: {got:?}");
        }
    }

    #[test]
    fn serialises_without_absent_fields() {
        let r = ProofRequest { v: 1, id: None, claim: "HoldsAtLeast".into(), zatoshi: 12_500_000, audience: "acme".into(), expiry_days: 7, respond_by: None, bind: None };
        let json = serde_json::to_string(&r).unwrap();
        assert_eq!(json, r#"{"v":1,"claim":"HoldsAtLeast","zatoshi":"12500000","audience":"acme","expiryDays":7}"#);
        assert_eq!(ProofRequest::decode(&pof_core::to_base64url(json.as_bytes())).unwrap(), r);
    }
}
